src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+package org.reactome.addlinks.dataretrieval.ensembl;
+
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.reactome.addlinks.dataretrieval.FileRetriever;
+
+import java.io.*;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.*;
+
+public class EnsemblBiomartRetriever extends FileRetriever {
+
+    public EnsemblBiomartRetriever() { }
+
+    public EnsemblBiomartRetriever(String retrieverName)
+    {
+        super(retrieverName);
+    }
+
+    private static final int UNIPROT_QUERY_TOTAL = 2;
+    private static int successCount;
+    private static final String  microarrayTypesBaseQuery = "type=listAttributes&mart=ENSEMBL_MART_ENSEMBL&virtualSchema=default&dataset=BIOMART_SPECIES_NAME_gene_ensembl&interface=default&attributePage=feature_page&attributeGroup=external&attributeCollection=microarray";
+
+    /**
+     * Downloads Ensembl-Microarray and Ensembl-Uniprot identifier mapping files for all species, if they exist.
+     * @throws Exception
+     */
+    protected void downloadData() throws Exception {
+
+        // TODO: Add unit tests
+        // Get names of all organisms we add links and/or microarray data for.
+        // Species names are in biomart format (eg: hsapiens).
+        for (String biomartSpeciesName : getBiomartSpeciesNames()) {
+            logger.info("Retrieving Biomart files for " + biomartSpeciesName);
+            // Create directory where Biomart files will be stored.
+            new File(this.destination).mkdirs();
+
+            logger.info("Retrieving microarray data");
+            // Query Biomart for existing microarray 'types' (not ids) that exist for this species.
+            Set<String> microarrayTypes = queryBiomart(getMicroarrayTypesQuery(biomartSpeciesName), 0);
+            // Iterate through each microarray type and retrieve Ensembl-Microarray identifier mappings.
+            // All mappings are stored in a single file, (eg: hsapiens_microarray);
+            retrieveAndStoreMicroarrayData(biomartSpeciesName, getBiomartXMLPath(), microarrayTypes);
+
+            // Query Ensembl-Uniprot (swissprot and trembl) identifier mapping data from Biomart
+            // and write it to a file (eg: hsapiens_uniprot).
+            logger.info("Retrieving UniProt data");
+            retrieveAndStoreUniprotData(biomartSpeciesName, getBiomartXMLPath());
+
+            logger.info("Completed Biomart data retrieval for " + biomartSpeciesName);
+        }
+    }
+
+    /**
+     * Query Biomart for both Uniprot mapping types (curated -- SwissProt, generated -- TrEMBL).
+     * @param biomartSpeciesName
+     * @throws Exception
```

Comment:
Nitpick alert :) - could descriptions be added for each JavaDoc annotation? 

It helps if/when we decide to create full JavaDoc pages for the projects and for submission to Sonatype.  For exceptions, I usually remove the `throws Exception` part of the method and let the IDE show me what is causing the exception to add a description of when the method throws.

Reply:
I'm a little confused by your last sentence. Do you mean removing the `throws Exception` from the method itself and surrounding the error-prone part with a try/catch? Or do you mean remove that temporarily so I can see what causes the exception (for example things related to MySQLAdaptor)? Or are you referring to the javadoc `@throws Exception`?

Reply:
It's a trick where you _temporarily_ remove `throws Exception` from the method declaration. Then the IDE should highlight all the lines inside the method that could thrown an exception, and the IDE should also hint to you what _kind_ of exception is thrown. with that information, you can then update the `throws` part of the method declaration (and also the javadoc) to throw more specific exceptions.

Reply:
Yeah, just removing the "throws Exception" part from the method declaration to see what gets highlighted as throwing the exception and then putting it back.  It shows you quickly what's causing the exception in the method to add info to the JavaDoc.